### PR TITLE
Add basic tagging

### PR DESCRIPTION
**Pull request by nefrathenrici**
Part of SDI #144 
### Content
This PR adds a new field `tags`, which will be a list of searchable tags that can be used to find the given parameter.
As an initial examples, I have added tags for the SurfaceFluxes.jl parameters.

In the API, these are supported by two new functions - 
- `get_tagged_parameter_names`: Returns a list of the parameters with a given tag.
- `get_tagged_parameter_values`: Returns a list of name-value Pairs of the parameters with a given tag.

For string-matching, punctuation and capitalization is removed. This could be changed to use string distance instead - I am not sure what the needs are.

I have also added tests for `get_tagged_parameter_values` and `get_tagged_parameter_names`

### CODE REVIEW

**Comment by trontrytel** [comment]
I will be able to tag my favorite parameters!

**Comment by nefrathenrici** [comment]
@trontrytel I am planning on adding tags for most parameters once the basic interface has been agreed upon!

**Comment by trontrytel** [comment]
> @trontrytel I am planning on adding tags for most parameters once the basic interface has been agreed upon!

Yes, this looks really cool. I think it will improve the readability a lot once fully implemented

File: docs/src/toml.md
Comment (odunbar) [inline docs/src/toml.md:39]: One should highlight in the comments that the search for tagging is done via this "fuzzy" matching, 

It may be useful to export fuzzy_match so users can test how their tags are interpreted in the current fuzzy_match function
e.g. currently it may not be clear to users that
```julia
fuzzy_match("Grachev", "gR*__aC.h,._e, --v-") # returns true
fuzzy_match("Grachev", "&gR*__aC.h,._e, --v-") # returns false
```

File: src/parameters.toml
Comment (odunbar) [inline src/parameters.toml:98]: If we put tags in the default file, be aware that this starts to set convention. If so, we should explicitly state the convention in the docs.

I can't quite make out the convention here. It seems to be a struct that will call this value (e.g. `Businger`) but also the components that build this struct `Businger`?), is that the idea?

File: docs/src/toml.md
Comment (odunbar) [inline docs/src/toml.md:43]: Highlight that tags are optional (this is not a namespace), and so the name must still be globally unique

File: docs/src/toml.md
Comment (odunbar) [inline docs/src/toml.md:50]: Perhaps be more explicit here, 

"If this convention is followed, to obtain the parameters used to build tagged by "Grachev" struct, one could call for example:
```julia 
# given toml_dict parsed from parameters.toml
grachev_params = get_tagged_parameter_values(toml_dict, "Grachev")
```

File: src/file_parsing.jl
Comment (odunbar) [inline src/file_parsing.jl:464]: Could this be a list of tags?

File: src/file_parsing.jl
Comment (odunbar) [inline src/file_parsing.jl:487]: could we do this for a list of tags

**Review by odunbar — COMMENTED** [review]
Thanks for this! 

1. I had a couple of overarching notes. Mainly, we should also add documentation for (i) how to check whether your tags are equal with fuzzy_match, (ii) how to call the get_tagged_... utilities, (iii) what the convention is for tagging default parameters. 

2. I wasn't expecting any tags to be applied in this PR, perhaps leave this to a future one?

File: src/parameters.toml
Comment (glwagner) [reply src/parameters.toml:98]: Maybe start conservatively with just the package it pertains to (`SurfaceFluxes`) in this case.

Comment (nefrathenrici) [reply src/parameters.toml:98]: I agree, it's best to start with just the package name for these parameters. We can determine conventions when we add tags for the rest of the defaults.

**Comment by nefrathenrici** [comment]
>     2. I wasn't expecting any tags to be applied in this PR, perhaps leave this to a future one?
@odunbar What do you think of just adding the "SurfaceFluxes" tag to the relevant parameters?

**Review by odunbar — COMMENTED** [review]
Thanks for the updates! I have one further point.
- Please change `tags` to `tag`. I notice that every other field is singular-named, even if it can take an item or vector-of-items as an value.

If it is convenient to add a tags then one can do this here. My concern is that changing these tags requires a PR (breaking if you remove tags) to this repo so make sure they are useful before they are added.

After this i will be happy for merge

**Comment by nefrathenrici** [comment]
@odunbar I have finally dealt with your last comments! I removed all of the tags from the default parameter file and renamed the `tags` field to `tag`.

File: docs/src/toml.md
Comment (odunbar) [inline docs/src/toml.md:15]: `tag`

**Review by odunbar — APPROVED** [review]
One `tags` slipped through the gap, also there are some functions missing from codecov could you add a quick test for `get_tagged_parameter_names` and `get_tagged_parameter_values` for the vector case?

Then merge away!

**Pull request merged**
